README.md
### CPUFreqStatusbar

Android Xposed module to show CPU frequency in the statusbar.

Based on [M1cha](https://github.com/M1cha)/[android_app_cputempinstatusbar](https://github.com/M1cha/android_app_cputempinstatusbar).


CHANGELOG
Version 1.1 - Sep 11, 2014
    * fixed preference initialization
    * added font size option
    * added vertical alignment option
    * text display tries to set a minimum (fixed) width, requires kernel cpufreq
      with scaling_available_frequencies or stats/time_in_state
    * doesn't run update when status bar is hidden (requested by aness zurba)
    * removed theme override
    * added custom icon
    * added logging option
    * misc small tweaks and clean up


src/org/coax/xposed/cpufreqstatusbar/Utils.java
package org.coax.xposed.cpufreqstatusbar;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

import android.content.Context;
import android.graphics.Paint;
import android.util.Log;
import de.robv.android.xposed.XposedBridge;

public class Utils {
	public static HashMap<String, Class<?>> prefs = new HashMap<String, Class<?>>();

	static {
		prefs.put("update_interval", String.class);
		prefs.put("position", String.class);
		prefs.put("frequency_file", String.class);
		prefs.put("measurement", String.class);
		prefs.put("manual_color", Boolean.class);
		prefs.put("color_mode", String.class);
		prefs.put("configured_color", Integer.class);
		prefs.put("color_low", Integer.class);
		prefs.put("color_middle", Integer.class);
		prefs.put("color_high", Integer.class);
		prefs.put("freq_middle", String.class);
		prefs.put("freq_high", String.class);
	}

	public static void log(String s) {
		XposedBridge.log("CPUFreqStatusbar: " + s);
	}
	
	public static String convertStreamToString(InputStream is) throws Exception {
	    BufferedReader reader = new BufferedReader(new InputStreamReader(is));
	    StringBuilder sb = new StringBuilder();
	    String line = null;
	    while ((line = reader.readLine()) != null) {
	      sb.append(line).append("\n");
	    }
	    return sb.toString();
	}
	
	public static String[] getFrequencyFiles() {
		ArrayList<String> result = new ArrayList<String>();
		result.add("Auto");
		
		try {
			InputStream in = Runtime.getRuntime()
					.exec("busybox find /sys -type f -name *cur_freq*")
					.getInputStream();
			BufferedReader inBuffered = new BufferedReader(
					new InputStreamReader(in));

			String line = null;
			while ((line = inBuffered.readLine()) != null) {
				result.add(line.trim());
			}
		} catch (IOException e) {
			e.printStackTrace();
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		return result.toArray(new String[]{});
	}
	
	private static String[] freqFiles = {
		"/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq",
		"/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_cur_freq",
	};
	
	public static File getFreqFile(Context context, String fileName) {
		File ret = null;
		
		if(fileName!=null) { 
			ret = new File(fileName);
			if(!ret.exists() || !ret.canRead())
				ret = null;
		}
		
		if(ret==null || fileName.equals("Auto")) {
			for(String freqFileName : freqFiles) {
				ret = new File(freqFileName);
				if(!ret.exists() || !ret.canRead()) {
					ret = null;
					continue;
				}
				else break;
			}
		}
		
		if(ret==null) {
			Utils.log("init could not find any frequency files");
		}
		
		return ret;
	}
	
	public static String findWidestFreqString() {
		File file = null;
		Scanner mScanner;
		String freq;
		String widestFreq = null;
		Paint mPaint = new Paint();

		float maxwidth = 0;
		float width;
		
		// size doesn't really matter here, it's the relative width that counts
		mPaint.setTextSize(16);

		file = new File("/sys/devices/system/cpu/cpu0/cpufreq/scaling_available_frequencies");
		if(file.exists() && file.canRead()) {
			// parse scaling_available_frequencies and find widest string 
			try {
				mScanner = new Scanner(file);
				while(mScanner.hasNext()) {
					freq = mScanner.next();
					freq = String.valueOf(Long.valueOf(freq)/1000);
					width = mPaint.measureText(freq);
					if(width > maxwidth) {
						widestFreq = freq;
						maxwidth = width;
					}
				}
				mScanner.close();
			} catch (FileNotFoundException e) {
				e.printStackTrace();
				Utils.log(Log.getStackTraceString(e));
			}
		} else {
			file = new File("/sys/devices/system/cpu/cpu0/cpufreq/stats/time_in_state");
			if(file.exists() && file.canRead()) {
				// parse time_in_state and find widest string 
				try {
					mScanner = new Scanner(file).useDelimiter("\\s+\\d+\n");
					while(mScanner.hasNext()) {
						freq = mScanner.next();
						freq = String.valueOf(Long.valueOf(freq)/1000);
						width = mPaint.measureText(freq);
						if(width > maxwidth) {
							widestFreq = freq;
							maxwidth = width;
						}
					}
					mScanner.close();
				} catch (FileNotFoundException e) {
					e.printStackTrace();
					Utils.log(Log.getStackTraceString(e));
				}
			}
		}
		
		return widestFreq;
	}
}


src/org/coax/xposed/cpufreqstatusbar/XposedInit.java
package org.coax.xposed.cpufreqstatusbar;

import java.lang.reflect.Method;
import java.util.ArrayList;
import info.mzimmermann.libxposed.LXTools;
import org.coax.xposed.cpufreqstatusbar.widget.CpuFreq;
import org.coax.xposed.cpufreqstatusbar.widget.LegacyPositionCallbackImpl;
import org.coax.xposed.cpufreqstatusbar.widget.PositionCallback;
import org.coax.xposed.cpufreqstatusbar.widget.PositionCallbackImpl;
import android.content.Context;
import android.util.Log;
import android.widget.TextView;
import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam;

public class XposedInit implements IXposedHookLoadPackage {
	private static ArrayList<TextView> tvClocks = new ArrayList<TextView>();
	private static PositionCallback mPositionCallback = null;
	private CpuFreq mCpuFreq = null;
	
	public static TextView getClock() {
		if(mPositionCallback==null) 
			return null;

		for(TextView tvClock : tvClocks) {
			if(mPositionCallback.getClockParent().findViewById(tvClock.getId())!=null) {
				return tvClock;
			}
		}

		Utils.log("init could not find clock object");
		return null;
	}
	
	public void handleLoadPackage(final LoadPackageParam lpparam)
			throws Throwable {
		if (!lpparam.packageName.equals("com.android.systemui"))
			return;
		
		XposedBridge.hookAllConstructors(XposedHelpers.findClass("com.android.systemui.statusbar.policy.Clock", lpparam.classLoader), new XC_MethodHook() {
			@Override
			protected void beforeHookedMethod(MethodHookParam param)
					throws Throwable {
				super.beforeHookedMethod(param);
				if(tvClocks.indexOf(param.thisObject)==-1) {
					tvClocks.add((TextView)param.thisObject);
				}
			}
		});
		
		// we hook this method to follow alpha changes in kitkat
		Method setAlpha = XposedHelpers.findMethodBestMatch(XposedHelpers.findClass("com.android.systemui.statusbar.policy.Clock", lpparam.classLoader), "setAlpha", Float.class);
		XposedBridge.hookMethod(setAlpha, new XC_MethodHook() {
			@Override
			protected void beforeHookedMethod(MethodHookParam param)
					throws Throwable {
				super.beforeHookedMethod(param);

				TextView mClock = XposedInit.getClock();
				if(param.thisObject!=mClock)
					return;

				if(mCpuFreq!=null && mClock!=null) {
					mCpuFreq.setAlpha(mClock.getAlpha());
				}
			}
		});

		XposedHelpers.findAndHookMethod(
			"com.android.systemui.statusbar.phone.PhoneStatusBar",
			lpparam.classLoader, "makeStatusBarView", new XC_MethodHook() {

				@Override
				protected void afterHookedMethod(MethodHookParam param)
						throws Throwable {
					try {
						// create cpufreq view
						Context mContext = (Context)XposedHelpers.getObjectField(param.thisObject, "mContext");
						LXTools.removeInvalidPreferences(Utils.prefs, mContext.getSharedPreferences(CpuFreq.PREF_KEY, 0));
						mCpuFreq = new CpuFreq(mContext);

						// identify legacy mode
						boolean legacy = false;
						try {
							XposedHelpers.getObjectField(param.thisObject, "mSystemIconArea");
						}
						catch(NoSuchFieldError e) {
							legacy = true;
						}
						
						// create callback
						if(legacy)
							mCpuFreq.mPositionCallback = new LegacyPositionCallbackImpl();
						else
							mCpuFreq.mPositionCallback = new PositionCallbackImpl();
						
						// initial setup
						mPositionCallback = mCpuFreq.mPositionCallback;
						mCpuFreq.mPositionCallback.setup(param, mCpuFreq);
						
						// set position
						LXTools.removeInvalidPreferences(Utils.prefs, mContext.getSharedPreferences(CpuFreq.PREF_KEY, 0));
						int position = Integer.parseInt(mContext.getSharedPreferences(CpuFreq.PREF_KEY, 0).getString("position", "0"));
						if(position==0) {
							mCpuFreq.mPositionCallback.setLeft();
						}
						else if(position==1) {
							mCpuFreq.mPositionCallback.setRight();
						}
						else if(position==2) {
							mCpuFreq.mPositionCallback.setAbsoluteLeft();
						}
					}
					catch(Exception e) {
						Utils.log(Log.getStackTraceString(e));
					}
				}
			}
		);
	}
}


src/org/coax/xposed/cpufreqstatusbar/activities/SettingsActivity.java
package org.coax.xposed.cpufreqstatusbar.activities;

import net.margaritov.preference.colorpicker.ColorPickerPreference;
import info.mzimmermann.libxposed.LXTools;
import info.mzimmermann.libxposed.apps.LXMyApp;
import org.coax.xposed.cpufreqstatusbar.R;
import org.coax.xposed.cpufreqstatusbar.Utils;
import org.coax.xposed.cpufreqstatusbar.widget.CpuFreq;
import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceActivity;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;

/**
 * A {@link PreferenceActivity} that presents a set of application settings. On
 * handset devices, settings are presented as a single list. On tablets,
 * settings are split by category, with category headers shown to the left of
 * the list of settings.
 * <p>
 * See <a href="http://developer.android.com/design/patterns/settings.html">
 * Android Design: Settings</a> for design guidelines and the <a
 * href="http://developer.android.com/guide/topics/ui/settings.html">Settings
 * API Guide</a> for more information on developing a Settings UI.
 */
public class SettingsActivity extends PreferenceActivity {
	/**
	 * Determines whether to always show the simplified settings UI, where
	 * settings are presented in a single list. When false, settings are shown
	 * as a master/detail two-pane view on tablets. When true, a single pane is
	 * shown on tablets.
	 */
	private static final boolean ALWAYS_SIMPLE_PREFS = false;
	public static final String ACTION_SETTINGS_UPDATE = "cpufreq-statusbar-settings-update";
	private static Context mContext = null;

	@Override
	public void onAttachedToWindow() {
		super.onAttachedToWindow();
		mContext = getApplicationContext();
		LXTools.removeInvalidPreferences(Utils.prefs, mContext.getSharedPreferences(CpuFreq.PREF_KEY, 0));
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);

		setupSimplePreferencesScreen();
	}

	/**
	 * Shows the simplified settings UI if the device configuration if the
	 * device configuration dictates that a simplified, single-pane UI should be
	 * shown.
	 */
	private void setupSimplePreferencesScreen() {
		if (!isSimplePreferences(this)) {
			return;
		}

		// In the simplified UI, fragments are not used at all and we instead
		// use the older PreferenceActivity APIs.

		// Add 'general' preferences.
		addPreferencesFromResource(R.xml.pref_general);

		ListPreference frequency_file = (ListPreference)findPreference("frequency_file");
		String[] files = Utils.getFrequencyFiles();
		frequency_file.setEntries(files);
		frequency_file.setEntryValues(files);
		if (frequency_file.getValue() == null) {
			frequency_file.setValue("Auto");
		}

		bindPreferenceSummaryToValue(findPreference("frequency_file"));
		bindPreferenceSummaryToValue(findPreference("update_interval"));
		bindPreferenceSummaryToValue(findPreference("position"));
		bindPreferenceSummaryToValue(findPreference("top_padding"));
		bindPreferenceSummaryToValue(findPreference("measurement"));
		bindPreferenceSummaryToValue(findPreference("show_unit"));
		bindPreferenceSummaryToValue(findPreference("font_size"));
		bindPreferenceSummaryToValue(findPreference("color_mode"));
		bindPreferenceSummaryToValue(findPreference("configured_color"));
		bindPreferenceSummaryToValue(findPreference("color_low"));
		bindPreferenceSummaryToValue(findPreference("color_middle"));
		bindPreferenceSummaryToValue(findPreference("color_high"));
		bindPreferenceSummaryToValue(findPreference("freq_middle"));
		bindPreferenceSummaryToValue(findPreference("freq_high"));
		bindPreferenceSummaryToValue(findPreference("enable_logging"));
	}

	/** {@inheritDoc} */
	@Override
	public boolean onIsMultiPane() {
		return isXLargeTablet(this) && !isSimplePreferences(this);
	}

	/**
	 * Helper method to determine if the device has an extra-large screen. For
	 * example, 10" tablets are extra-large.
	 */
	private static boolean isXLargeTablet(Context context) {
		return false;
	}

	/**
	 * Determines whether the simplified settings UI should be shown. This is
	 * true if this is forced via {@link #ALWAYS_SIMPLE_PREFS}, or the device
	 * doesn't have newer APIs like {@link PreferenceFragment}, or the device
	 * doesn't have an extra-large screen. In these cases, a single-pane
	 * "simplified" settings UI should be shown.
	 */
	private static boolean isSimplePreferences(Context context) {
		return ALWAYS_SIMPLE_PREFS
				|| Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB
				|| !isXLargeTablet(context);
	}

	/**
	 * A preference value change listener that updates the preference's summary
	 * to reflect its new value.
	 */
	private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
		@Override
		public boolean onPreferenceChange(Preference preference, Object value) {
			String stringValue = value==null ? "" : value.toString();
			PreferenceManager pm = preference.getPreferenceManager();

			// get color_mode
			String sColorMode = "0";
			int color_mode = 0;
			try {
				if(preference.getKey().equals("color_mode"))
					sColorMode = stringValue;
				else
					sColorMode = PreferenceManager.getDefaultSharedPreferences(
							preference.getContext()).getString("color_mode", "0");
				color_mode = Integer.parseInt(sColorMode);
			}
			catch(Exception e){
			}

			switch(color_mode) {
			case 1:
				pm.findPreference("configured_color").setEnabled(true);
				pm.findPreference("color_low").setEnabled(false);
				pm.findPreference("color_middle").setEnabled(false);
				pm.findPreference("color_high").setEnabled(false);
				pm.findPreference("freq_middle").setEnabled(false);
				pm.findPreference("freq_high").setEnabled(false);
				break;
			case 2:
				pm.findPreference("configured_color").setEnabled(false);
				pm.findPreference("color_low").setEnabled(true);
				pm.findPreference("color_middle").setEnabled(true);
				pm.findPreference("color_high").setEnabled(true);
				pm.findPreference("freq_middle").setEnabled(true);
				pm.findPreference("freq_high").setEnabled(true);
				break;
			case 0:
			default:
				pm.findPreference("configured_color").setEnabled(false);
				pm.findPreference("color_low").setEnabled(false);
				pm.findPreference("color_middle").setEnabled(false);
				pm.findPreference("color_high").setEnabled(false);
				pm.findPreference("freq_middle").setEnabled(false);
				pm.findPreference("freq_high").setEnabled(false);
				break;
			}

			if (preference instanceof ListPreference) {
				// For list preferences, look up the correct display value in
				// the preference's 'entries' list.
				ListPreference listPreference = (ListPreference) preference;
				int index = listPreference.findIndexOfValue(stringValue);

				// Set the summary to reflect the new value.
				preference.setSummary(index >= 0 ? listPreference.getEntries()[index] : null);
			} else if(!(preference instanceof ColorPickerPreference) && !(preference instanceof CheckBoxPreference)) {
				// For all other preferences, set the summary to the value's
				// simple string representation.
				preference.setSummary(stringValue);
			}
			return true;
		}
	};

	/**
	 * Binds a preference's summary to its value. More specifically, when the
	 * preference's value is changed, its summary (line of text below the
	 * preference title) is updated to reflect the value. The summary is also
	 * immediately updated upon calling this method. The exact display format is
	 * dependent on the type of preference.
	 * 
	 * @see #sBindPreferenceSummaryToValueListener
	 */
	private static void bindPreferenceSummaryToValue(Preference preference) {
		// Set the listener to watch for value changes.
		LXMyApp.setTransferOnPreferenceChangeListener(ACTION_SETTINGS_UPDATE, preference, sBindPreferenceSummaryToValueListener);

		// Trigger the listener immediately with the preference's
		// current value.
		sBindPreferenceSummaryToValueListener.onPreferenceChange(
				preference,
				PreferenceManager.getDefaultSharedPreferences(preference.getContext())
					.getAll().get(preference.getKey()));
	}

	/**
	 * This fragment shows general preferences only. It is used when the
	 * activity is showing a two-pane settings UI.
	 */
	@TargetApi(Build.VERSION_CODES.HONEYCOMB)
	public static class GeneralPreferenceFragment extends PreferenceFragment {
		@Override
		public void onCreate(Bundle savedInstanceState) {
			super.onCreate(savedInstanceState);

			addPreferencesFromResource(R.xml.pref_general);
		}
	}
}


src/org/coax/xposed/cpufreqstatusbar/widget/PositionCallbackImpl.java
package org.coax.xposed.cpufreqstatusbar.widget;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.widget.LinearLayout;
import de.robv.android.xposed.XC_MethodHook.MethodHookParam;
import de.robv.android.xposed.XposedHelpers;

public class PositionCallbackImpl implements PositionCallback {
	
	private LinearLayout mSystemIconArea;
	private LinearLayout mStatusBarContents;
	private LinearLayout container;
	private View cpufreq;

	@Override
	public void setup(MethodHookParam param, View v) {
		 cpufreq = v;
		 mSystemIconArea = (LinearLayout)XposedHelpers.getObjectField(param.thisObject, "mSystemIconArea");
         mStatusBarContents = (LinearLayout)XposedHelpers.getObjectField(param.thisObject, "mStatusBarContents");
         Context mContext = (Context)XposedHelpers.getObjectField(param.thisObject, "mContext");

         container = new LinearLayout(mContext);
         container.setOrientation(LinearLayout.HORIZONTAL);
         container.setWeightSum(1);
         container.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.MATCH_PARENT));
         container.setVisibility(View.GONE);
         mStatusBarContents.addView(container, 0);
	}

	@Override
	public void setAbsoluteLeft() {
		mSystemIconArea.removeView(cpufreq);
		container.removeView(cpufreq);

		container.addView(cpufreq);
		container.setVisibility(View.VISIBLE);
	}

	@Override
	public void setLeft() {
		mSystemIconArea.removeView(cpufreq);
		container.removeView(cpufreq);

        mSystemIconArea.addView(cpufreq, 0);
        container.setVisibility(View.GONE);
	}

	@Override
	public void setRight() {
		mSystemIconArea.removeView(cpufreq);
		container.removeView(cpufreq);

        mSystemIconArea.addView(cpufreq);
        container.setVisibility(View.GONE);
	}

	@Override
	public LinearLayout getClockParent() {
		return mStatusBarContents;
		
	}
}


src/org/coax/xposed/cpufreqstatusbar/widget/LegacyPositionCallbackImpl.java
package org.coax.xposed.cpufreqstatusbar.widget;

import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.XC_MethodHook.MethodHookParam;

public class LegacyPositionCallbackImpl implements PositionCallback {

	private LinearLayout mStatusIcons;
	private LinearLayout mIcons;
	private View cpufreq;
	private LinearLayout mNotificationIconArea;

	@Override
	public void setup(MethodHookParam param, View v) {
		 cpufreq = v;
		 mStatusIcons = (LinearLayout)XposedHelpers.getObjectField(param.thisObject, "mStatusIcons");
		 mIcons = (LinearLayout)XposedHelpers.getObjectField(param.thisObject, "mIcons");
		 mNotificationIconArea = (LinearLayout)mIcons.getChildAt(0);
	}

	@Override
	public void setAbsoluteLeft() {
		removeFromParent();
		mNotificationIconArea.addView(cpufreq, 0);
	}

	@Override
	public void setLeft() {
		removeFromParent();
		mIcons.addView(cpufreq, mIcons.indexOfChild(mStatusIcons));
	}

	@Override
	public void setRight() {
		removeFromParent();
		mIcons.addView(cpufreq);
	}
	
	private void removeFromParent() {
		if(cpufreq.getParent()!=null)
			((ViewGroup)cpufreq.getParent()).removeView(cpufreq);
	}

	@Override
	public LinearLayout getClockParent() {
		return mIcons;
	}
}


src/org/coax/xposed/cpufreqstatusbar/widget/CpuFreq.java
package org.coax.xposed.cpufreqstatusbar.widget;

import info.mzimmermann.libxposed.LXTarget;
import org.coax.xposed.cpufreqstatusbar.Utils;
import org.coax.xposed.cpufreqstatusbar.XposedInit;
import org.coax.xposed.cpufreqstatusbar.activities.SettingsActivity;
import java.io.File;
import java.io.FileInputStream;
import android.annotation.SuppressLint;
import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.graphics.Color;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.TextView;

@SuppressLint("HandlerLeak")
public class CpuFreq extends TextView implements OnSharedPreferenceChangeListener {
	final public static String INTENT_ACTION_UPDATE = "cpufreq_update_timer";
	final public static String PREF_KEY = "org.coax.xposed.cpufreqstatusbar_preferences";
	final private Context mContext;
	private PendingIntent pi = null;
	private File freqFile = null;

	public PositionCallback mPositionCallback = null;

	private boolean isScreenOn = true;
	private boolean isStatusBarVis = true;

	private String sWidestFreq;
	private int statusbarHeight = 0;
	private final static int paddingWidth = 3;
	
	private boolean loggingEnabled = false;

	public CpuFreq(Context context) {
		this(context, null);
	}

	public CpuFreq(Context context, AttributeSet attrs) {
		this(context, attrs, 0);
	}

	public CpuFreq(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
		
		mContext = context;
		
		// init
		loggingEnabled = mContext.getSharedPreferences(PREF_KEY, 0).getBoolean("enable_logging", false);

		freqFile = Utils.getFreqFile(mContext, mContext.getSharedPreferences(PREF_KEY, 0).getString("frequency_file", null));
		if(loggingEnabled)
			Utils.log("init freqFile = " + (freqFile == null ? "null" : freqFile.getPath()));

		// set height
		setLayoutParams(new ViewGroup.LayoutParams(
				ViewGroup.LayoutParams.WRAP_CONTENT,
				ViewGroup.LayoutParams.MATCH_PARENT));
		int resourceId = getResources().getIdentifier("status_bar_height", "dimen", "android");
		if(resourceId > 0)
			statusbarHeight = getResources().getDimensionPixelSize(resourceId);
		if(statusbarHeight > 0) {
			setHeight(statusbarHeight);
		}
		if(loggingEnabled)
			Utils.log("init statusbarHeight = " + statusbarHeight);

		// set fixed width
		sWidestFreq = Utils.findWidestFreqString();
		if(sWidestFreq != null) {
			String sFormattedWidestFreq;
			String measurement = mContext.getSharedPreferences(PREF_KEY, 0).getString("measurement", "M");
			if(measurement.equals("G")){
				float fWidestFreq = Float.parseFloat(sWidestFreq)/1000F;
				sFormattedWidestFreq = String.format("%.2f", fWidestFreq);
			} else {
				sFormattedWidestFreq = sWidestFreq;
			}
			if(mContext.getSharedPreferences(PREF_KEY, 0).getBoolean("show_unit", false)) {
				sFormattedWidestFreq = sFormattedWidestFreq + measurement;
			}
			setMinimumWidth((int)getPaint().measureText(sFormattedWidestFreq) + paddingWidth * 2);
			if(loggingEnabled) {
				Utils.log("init sFormattedWidestFreq = " + sFormattedWidestFreq);
				Utils.log("init setMinimumWidth = " + getPaint().measureText(sFormattedWidestFreq));
			}
		} else if(loggingEnabled) {
			Utils.log("init could not set width");
		}

		// style
		setTextColor(Color.WHITE);
		setTextSize(TypedValue.COMPLEX_UNIT_SP,
					Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("font_size", "16")));
		setPadding(paddingWidth,
					Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("top_padding", "2")),
					paddingWidth,
					0);
		setGravity(Gravity.TOP | Gravity.RIGHT);  // gravity goes to TOP for manual top_padding
	}
	
	@Override
	protected void onAttachedToWindow() {
		super.onAttachedToWindow();

		IntentFilter filter = new IntentFilter();
		filter.addAction(INTENT_ACTION_UPDATE);
		filter.addAction(SettingsActivity.ACTION_SETTINGS_UPDATE);
		filter.addAction(Intent.ACTION_SCREEN_OFF);
		filter.addAction(Intent.ACTION_SCREEN_ON);
		mContext.registerReceiver(mBroadcastReceiver, filter);

		mContext.getSharedPreferences(PREF_KEY, 0).registerOnSharedPreferenceChangeListener(this);

		setOnSystemUiVisibilityChangeListener(new OnSystemUiVisibilityChangeListener() {
			@Override
			public void onSystemUiVisibilityChange(int visibility) {
				if((visibility & SYSTEM_UI_FLAG_FULLSCREEN) > 0) {
					isStatusBarVis = false;
				} else {
					isStatusBarVis = true;
				}
			}
		});
		
		// start update interval
		int updateInterval = Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("update_interval", "1000"));
		setAlarm(updateInterval);
	}
	
	@Override
	protected void onDetachedFromWindow() {
		mContext.unregisterReceiver(mBroadcastReceiver);
		mContext.getSharedPreferences(PREF_KEY, 0).unregisterOnSharedPreferenceChangeListener(this);
		cancelAlarm();
		super.onDetachedFromWindow();
	}

	private final BroadcastReceiver mBroadcastReceiver = new BroadcastReceiver() {
		@Override
		public void onReceive(Context context, Intent intent) {
			if(intent.getAction().equals(Intent.ACTION_SCREEN_ON)) {
				isScreenOn = true;
			}
			else if(intent.getAction().equals(Intent.ACTION_SCREEN_OFF)) {
				isScreenOn = false;
			}
			else if(intent.getAction().equals(INTENT_ACTION_UPDATE) && isScreenOn && isStatusBarVis) {
				updateFrequency();
			}
			else if(intent.getAction().equals(SettingsActivity.ACTION_SETTINGS_UPDATE)) {
				LXTarget.receivePreferences(context.getSharedPreferences(PREF_KEY, 0), intent);
			}
		}
	};

	public void setAlarm(int interval) {
		AlarmManager am = (AlarmManager) mContext
				.getSystemService(Context.ALARM_SERVICE);
		Intent intent = new Intent(INTENT_ACTION_UPDATE);
		pi = PendingIntent.getBroadcast(mContext, 0, intent, 0);
		am.setRepeating(AlarmManager.RTC, System.currentTimeMillis(), interval, pi);
	}
	
	public void cancelAlarm() {
		AlarmManager am = (AlarmManager) mContext
				.getSystemService(Context.ALARM_SERVICE);
		if(pi!=null)  {
			am.cancel(pi);
			pi.cancel();
		}
	}

	private void updateFrequency() {
		try {
			FileInputStream fis = new FileInputStream(freqFile);
			StringBuffer sbFreq = new StringBuffer("");

			// read frequency
			byte[] buffer = new byte[1024];
			while(fis.read(buffer) != -1) {
				sbFreq.append(new String(buffer));
			}
			fis.close();

			// get freq value and convert to MHz
			String sFreq = sbFreq.toString().replaceAll("[^0-9]+", "");
			Long lFreq = Long.valueOf(sFreq)/1000;
			int iFreq = lFreq.intValue();
			
			// measure system
			String measurement = mContext.getSharedPreferences(PREF_KEY, 0).getString("measurement", "M");
			if(measurement.equals("M")){
				sFreq = String.valueOf(iFreq);
			}
			else if(measurement.equals("G")) {
				float fFreq = iFreq/1000F;
				sFreq = String.format("%.2f", fFreq);
			}

			// unit label
			boolean show_unit = mContext.getSharedPreferences(PREF_KEY, 0).getBoolean("show_unit", false);
			if(show_unit) {
				sFreq = sFreq + measurement;
			}
			
			// set text
			setText(sFreq);
			
			// set text color
			int color_mode = Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("color_mode", "0"));
			TextView mClock = XposedInit.getClock();
			
			// set alpha, this prevents wrong initial colors
			if(mClock != null) {
				setAlpha(mClock.getAlpha());
			}

			switch(color_mode) {
			// auto
			case 0:
				if(mClock != null) {
					setTextColor(mClock.getCurrentTextColor());
				}
				break;
			//manual
			case 1:
				int configured_color = mContext.getSharedPreferences(PREF_KEY, 0).getInt("configured_color", Color.WHITE);
				setTextColor(configured_color);
				break;
			//freq 
			case 2:
				int color_low = mContext.getSharedPreferences(PREF_KEY, 0).getInt("color_low", Color.GREEN);
				int color_middle = mContext.getSharedPreferences(PREF_KEY, 0).getInt("color_middle", Color.YELLOW);
				int color_high = mContext.getSharedPreferences(PREF_KEY, 0).getInt("color_high", Color.RED);
				int freq_middle = Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("freq_middle", "500"));
				int freq_high = Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("freq_high", "1000"));
				
				if(iFreq >= freq_high)
					setTextColor(color_high);
				else if(iFreq >= freq_middle)
					setTextColor(color_middle);
				else
					setTextColor(color_low);
				break;
			}
		} catch (Exception e) {
			e.printStackTrace();
			Utils.log(Log.getStackTraceString(e));
			setText("-");
		}
	}

	@Override
	public void onSharedPreferenceChanged(SharedPreferences pref, String key) {
		if(key.equals("update_interval")) {
			int updateInterval = Integer.parseInt(pref.getString("update_interval", "1000"));
			cancelAlarm();
			setAlarm(updateInterval);
		}
		
		else if(key.equals("frequency_file")) {
			String frequency_file = pref.getString("frequency_file", null);
			freqFile = Utils.getFreqFile(mContext, frequency_file);
		}

		else if(key.equals("position")) {
			int position = Integer.parseInt(pref.getString("position", "0"));
			if(getParent()!=null)
				((ViewGroup)getParent()).removeView(this);
			if(position == 0) {
				mPositionCallback.setLeft();
			}
			else if(position == 1) {
				mPositionCallback.setRight();
			}
			else if(position == 2) {
				mPositionCallback.setAbsoluteLeft();
			}
		}

		else if(key.equals("font_size") || key.equals("top_padding") ||
				key.equals("measurement") || key.equals("show_unit")) {
			// set font size
			setTextSize(TypedValue.COMPLEX_UNIT_SP,
					Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("font_size", "16")));
			// set padding
			setPadding(paddingWidth,
					Integer.parseInt(mContext.getSharedPreferences(PREF_KEY, 0).getString("top_padding", "2")),
					paddingWidth,
					0);
			// set fixed width
			if(sWidestFreq != null) {
				String sFormattedWidestFreq;
				String measurement = pref.getString("measurement", "M");
				if(measurement.equals("G")){
					float fWidestFreq = Float.parseFloat(sWidestFreq)/1000F;
					sFormattedWidestFreq = String.format("%.2f", fWidestFreq);
				} else {
					sFormattedWidestFreq = sWidestFreq;
				}
				if(pref.getBoolean("show_unit", false)) {
					sFormattedWidestFreq = sFormattedWidestFreq + measurement;
				}
				setMinimumWidth((int)getPaint().measureText(sFormattedWidestFreq) + paddingWidth * 2);
				if(loggingEnabled) {
					Utils.log("pref sFormattedWidestFreq = " + sFormattedWidestFreq);
					Utils.log("pref setMinimumWidth = " + getPaint().measureText(sFormattedWidestFreq));
				}
			}
		}

		else if(key.equals("enable_logging")) {
			loggingEnabled = mContext.getSharedPreferences(PREF_KEY, 0).getBoolean("enable_logging", false);
		}

		updateFrequency();
	}
}
